Fix truediv numerics between eager and compile

PaulZhang12 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164144

Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

PaulZhang12 committed:
Fix truediv numerics between eager and compile

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1053,7 +1053,10 @@ def abs(x):
     @staticmethod
     def truediv(x, y):
-        out = f"({x} / {y})"
+        if x.dtype == torch.float32 and y.dtype == torch.float32:
```
I am not sure.. I can do a followup and check

PaulZhang12 replied:
https://www.internalfb.com/code/fbsource/[5291bf823ebee20df47c5b01645639a1f93da5ab]/fbcode/caffe2/caffe2/operators/elementwise_div_op.h?lines=21 I believe it's here? Lots of different layers to go through

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 commented:
@ngimel the test case in this pr exposes the numerical differences, even with converting both to a tensor and doing tensor / tensor in Inductor. Am I misunderstanding your point?

ngimel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1062,7 +1062,13 @@ def abs(x):
+        if x_dtype == torch.float32 and y_dtype == torch.float32:
+            out = f"triton.language.div_rn({x}, {y})"
+        else:
+            out = f"({x} / {y})"
```
can you leave a comment that `x /y` in triton is lowered to `dif.full` which is approximate, instead of `div_rn`

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix truediv numerics between eager and compile"


Addresses numeric differences between eager and compile in https://github.com/pytorch/pytorch/issues/141753



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]
